File: 3433-count-mentions-per-user/3433-count-mentions-per-user.cpp
class Solution {
public:
    void applymsg(vector<string>&event , vector<int>&mentions , vector<int>&offline){
        int time = stoi(event[1]);
        vector<string> ids;
        stringstream ss(event[2]);

        string token;
        while(ss >> token){
            ids.push_back(token);
        }

        for(string id: ids){
            if(id=="ALL"){
                for(int i=0;i<mentions.size();i++){
                    mentions[i]++;
                }
            }
            else if(id=="HERE"){
                for(int i=0;i<mentions.size();i++){
                    if(offline[i]==0 || offline[i]+60 <= time){
                        mentions[i]++;
                    }
                }
            }
            else{
                mentions[stoi(id.substr(2))]++;
            }
        }

    }


    vector<int> countMentions(int n, vector<vector<string>>& events) {
        auto lambda = [](auto&v1 , auto&v2){
            int t1= stoi(v1[1]);
            int t2= stoi(v2[1]);
            if(t1==t2){
                return v1[0]>v2[0];
            }
            return t1<t2;
        };
        sort(events.begin(), events.end(), lambda);

        vector<int>mentions(n);
        vector<int>offline(n);
         
        for(auto&event : events){
            if(event[0]=="MESSAGE"){
                applymsg(event , mentions , offline);
            }
            else if(event[0]=="OFFLINE"){
                int time = stoi(event[1]);
                int id= stoi(event[2]);
                offline[id] = time;
            }
        }
        return mentions;
    }
};

File: 3433-count-mentions-per-user/3433-count-mentions-per-user.java
import java.util.*;

class Solution {
    void applyMsg(List<String> event, int[] mentions, int[] offline) {
        int time = Integer.parseInt(event.get(1));
        String[] tokens = event.get(2).split(" ");
        for (String id : tokens) {
            if (id.equals("ALL")) {
                for (int i = 0; i < mentions.length; i++) {
                    mentions[i]++;
                }
            } 
            else if (id.equals("HERE")) {
                for (int i = 0; i < mentions.length; i++) {
                    // offline[i] == -1 means "never offline"
                    if (offline[i] == -1 || offline[i] + 60 <= time) {
                        mentions[i]++;
                    }
                }
            } 
            else {
                int idx = Integer.parseInt(id.substring(2));
                mentions[idx]++;
            }
        }
    }

    public int[] countMentions(int n, List<List<String>> events) {

        // sort by time, then type lexicographically
        events.sort((v1, v2) -> {
            int t1 = Integer.parseInt(v1.get(1));
            int t2 = Integer.parseInt(v2.get(1));
            if (t1 != t2) return t1 - t2;
            return v2.get(0).compareTo(v1.get(0));
        });

        int[] mentions = new int[n];
        int[] offline = new int[n];

        // initialize offline times to -1 (never offline)
        Arrays.fill(offline, -1);

        for (List<String> event : events) {
            String type = event.get(0);

            if (type.equals("MESSAGE")) {
                applyMsg(event, mentions, offline);
            } 
            else if (type.equals("OFFLINE")) {
                int time = Integer.parseInt(event.get(1));
                int id = Integer.parseInt(event.get(2));
                offline[id] = time;
            }
        }

        return mentions;
    }
}
